src/target/xtensa/xtensa.c

```diff
@@ -1278,51 +1439,52 @@ int xtensa_resume(struct target *target,
 
 static bool xtensa_pc_in_winexc(struct target *target, target_addr_t pc)
 {
+	struct xtensa *xtensa = target_to_xtensa(target);
 	uint8_t insn_buf[XT_ISNS_SZ_MAX];
 	int err = xtensa_read_buffer(target, pc, sizeof(insn_buf), insn_buf);
 	if (err != ERROR_OK)
 		return false;
 
 	xtensa_insn_t insn = buf_get_u32(insn_buf, 0, 24);
-	xtensa_insn_t masked = insn & XT_INS_L32E_S32E_MASK;
-	if (masked == XT_INS_L32E(0, 0, 0) || masked == XT_INS_S32E(0, 0, 0))
+	xtensa_insn_t masked = insn & XT_INS_L32E_S32E_MASK(xtensa);
+	if (masked == XT_INS_L32E(xtensa, 0, 0, 0) || masked == XT_INS_S32E(xtensa, 0, 0, 0))
 		return true;
 
-	masked = insn & XT_INS_RFWO_RFWU_MASK;
-	if (masked == XT_INS_RFWO || masked == XT_INS_RFWU)
+	masked = insn & XT_INS_RFWO_RFWU_MASK(xtensa);
+	if (masked == XT_INS_RFWO(xtensa) || masked == XT_INS_RFWU(xtensa))
 		return true;
 
 	return false;
 }
 
+/* TODO check PS change flow in the upstream. Looks like upstream branch has some issue. This is working version. */
```

Comment:
@ianstcdns fyi

Reply:
was this the issue we fixed upstream with the PS vs. EPSn change?  if not, what's happening, and is there a way to reproduce it on mainline (either with my wrover-kit or a generic xtensa target)?

Reply:
IIRC, there was a difference between our implementations related to $icountlevel.  LX uses this register to determine which instructions "count" during instruction stepping...e.g. if an instruction is run when CINTLEVEL >= ICOUNTLEVEL, then it does not increment $icount.  This allows OOCD to set $icountlevel to 1 and single-step user code, and if an overflow exception or L1 interrupt comes in, allow it to run until the core returns to level 0 and $icount increments up to 0.  So, we're not masking interrupts (they still occur), but we also don't step into them.

Forcibly disabling interrupts during step can cause unintended side-effects in the system, such as missing timer ticks or watchdog kicks, and should really be avoided.  

I think I may have changed the behavior of "xtensa maskisr" to behave more as a "step into ISR" vs. "step over ISR" flag that leave the ISR active in both cases.  Could your test be assuming the ISR is getting disabled instead?

Reply:
Ok I found the old discussion. https://github.com/erhankur/review-openocd/pull/3#discussion_r901551388

Reply:
This is the test  we are talking about. 
https://github.com/erhankur/openocd-esp32/blob/master/testing/esp/test_step.py#L363
https://github.com/erhankur/openocd-esp32/blob/master/testing/esp/test_apps/gen_ut_app/main/gen_ut_app.c#L307